### DISCUSSION TITLE
git query 'stack()' returns forks in the stack

### DISCUSSION THREAD

**Discussion by pink** [arxanas/git-branchless#958] [Q&A] [answered]
When there's a fork in the graph, is the expected behavior of `git query 'stack()' -b` still [this](https://github.com/arxanas/git-branchless/blob/2b581a7af588458bcfb380d879c46b44af6881da/git-branchless/src/revset/eval.rs#L175-L186)?

```
// Note that for a graph like
//
// ```
// O
// |
// o A
// | \
// |  o B
// |
// @ C
// ```
// this will return `{A, B, C}`, not just `{A, C}`.
```

and if so, is there a way to accomplish getting the latter? For more context, my use case is constructing a stack of branches that I can eventually push to Github, where the chain would look something like `master <- branch_a <- branch_c`. Using the current ordering of this query would result in `master <- branch_a <- branch_b <- branch_c`, which isn't correct.

**Comment by arxanas** [comment]
This is intentional. I often use `stack():@`, which is all commits in the stack which are reachable from the current `HEAD` (`@`).

**Reply by pink** [reply]
Ahh interesting. Expanding on the previous example, if A, B, and C were each branches, and there were multiple commits between A and B, it seems like `git query 'stack():@' -b ` would exclude branch B from the results if run in between the two.

```
// ```
// O
// |
// o A
// | \
// |  @
// |   \
// |    o B
// |
// o C
// ```
```

Is there a way to show commits/branches reachable in either direction?

**Reply by arxanas** [reply, marked as answer]
If I'm understanding the question correctly, you should be able to get both sides with `stack():@ | @:stack()`, does that work?

**Reply by pink** [reply]
Oh nice, yea I think that pretty much works! One related question though - do you know if it's possible to get the closest descendant branch from a commit using `git query`?

I'm trying to create an alias for only pushing the current branch and its ancestors. Similar to the previous example, the case I'm trying to resolve is when `HEAD` is in between branches - `@:stack()` would give me every descendant branch but ideally I'd like to just get the "current" one. I think I can solve this with some scripting instead, but just curious if there's a way to express this with the existing functions.

**Reply by claytonrcarter** [reply]
If you know you have a non-forking (linear) graph between `@` and the next branch, then something like this may work: `roots( ( @:stack() & branches() ): )`

1. `@:stack()` - you know this one
2. `... & branches()` - limit to commits with branch pointers
3. `(...):` - all descendants of those branch commits
4. `roots(...)` - only the root ancestors of the set
5. you could even wrap this in `exactly(1, ...)` so that it will only succeed if there is 1 matching commit, in case you *do* call this on a stack w/ a non-linear descendant graph

So I guess your alias may be `git submit -c 'exactly(1, roots( ( @:stack() & branches() ): ))` ... revsets FTW?!

**Reply by arxanas** [reply]
You can also write `descendants(X)` instead of `X:` if you find it more readable (when `X` is a big expression).

The function `roots(descendants(X))` is common enough that I think it ought to be a built-in function (and likewise for `heads(ancestors(X))`).

**Reply by pink** [reply]
@claytonrcarter Whoa that's slick, thanks for the explanation! Spent a while today trying to formulate a query like this - really appreciate both of your help 😄.
